src/main/java/org/jabref/gui/StateManager.java

```diff
@@ -112,4 +118,20 @@ public void setSearchQuery(SearchQuery searchQuery) {
     public OptionalObjectProperty<Node> focusOwnerProperty() { return focusOwner; }
 
     public Optional<Node> getFocusOwner() { return focusOwner.get(); }
+
+    public UiThreadObservableList<Task> getBackgroundTasks() {
+        return backgroundTasks;
+    }
+
+    public void addBackgroundTask(Task backgroundTask) {
+        this.backgroundTasks.add(backgroundTask);
+    }
+
+    public BooleanBinding anyTaskRunningBinding = Bindings.createBooleanBinding(
```

Comment:
For the bindings to update, you need to add the list as a dependency (second argument of the `createXBinding` method). In the case of lists, however, its easier to use `EasyBind.combine`: https://github.com/TomasMikula/EasyBind#combine-list

Reply:
Aha, that looks useful.
However, I think I am running into some typing issues which I am struggling to resolve.

With the library you pointed me to, I wound up with the following:

`
    public Binding<Boolean> anyTaskRunningBinding = EasyBind.combine(
            backgroundTasks,
            stream -> stream.anyMatch(Task::getProgress)
    );
`

This gives me an error that it expects a Binding<Boolean>, but gets a MonadicBinding<R>.
If I just cast it, I get the following:

`no instance(s) of type variable(s) T exist so that Task conforms to ObservableValue<? extends T>`

I think this is the issue I need to resolve first. As the example in the library works fine, I guess the conversion from MonadicBinding<R> to Binding<Boolean> is then done implicitly, is that correct?

I struggle solving this because I dont know the type parameter of the tasks I am storing.
When storing, the Task has type V:

https://github.com/btut/jabref/blob/b8834914e5e735abaa7a710216888abe6bf54277/src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java#L99-L104

But in the list I just use Task:

https://github.com/btut/jabref/blob/b8834914e5e735abaa7a710216888abe6bf54277/src/main/java/org/jabref/gui/StateManager.java#L49

How do I need to change the list in order for it to work?

Reply:
Add the generics to the Task as welll:
`    private final UiThreadObservableList<Task<V>> backgroundTasks = new UiThreadObservableList(FXCollections.observableArrayList());`

Reply:
But where would I get the V from in this case?
`Cannot resolve symbol 'V'`

Reply:
I changed it to Task<?> now and now the progress, title and message properties make their way through to the dialogue.
I still cannot create the bindings for the progress indicator though.

Reply:
That change also allows me to use EasyBind's listBind to bind the task list in the view to the task list in StateManager.

Reply:
![image](https://user-images.githubusercontent.com/10479048/81391853-14aab680-911e-11ea-98a6-b63c1935a3df.png)
Here is the first peak at a download in the dialogue (which will probably be changed to a popover later). As you can see, the download task has it's title and message set and progress is updated fine. However, there are still a lot of tasks that do not have any details.

Reply:
Looks good already!
There are a lot of tasks which don't concern file downloads and just perform some operations in the background. If you search for all references from backgrond task you probably have to adjust each one to add a meaninful property

Reply:
I think that should be doable. But before I do that: is it a good idea to have all background tasks listed? Or should we just show downloads? If we go with the first option, I suggest turning off the retain-tasks feature of the task view. Then, all completed tasks will automatically disappear. This way we have less tasks that only ran for a very little time filling up the view.

I got the bindings to work by storing a list of Property<Task<?>> Instead of Task<?>. That does the trick for the progress indicator. It now is indeterminate when one of the tasks has an indeterminate progress and shows the average progress otherwise (100% if no tasks are running).
For some reason, this breaks the task view. Since I now store a list of properties, and not a list of tasks, I cannot bind them directly to the view, so I went back to doing it manually, but that does not seem to work.

https://github.com/btut/jabref/blob/38dd89dce28b72d195de18b848b011532ef1f868/src/main/java/org/jabref/gui/taskprogressmanager/TaskProgressDialog.java#L34-L47

Reply:
Now I just converted the list of properties back into a list of tasks using EasyBind and the bind that list to the list of tasks in the view (again using EasyBind). Now both the indicator and the dialogue work fine and are updated with the running tasks!